### ISSUE TITLE
Adding authenicationGuard to route is not working as expected

### ISSUE THREAD

**Issue by bikramchoudhury** [creativetimofficial/material-dashboard-pro-angular#84]
Hi Team,
I have changed the app route configuration to make my login page as a landing page. And I have added authenitcationGuard for the admin layout (ex: dashboard).
Without login, it's redirecting to the login page as expected, but after login, it's not going to admin layout (ex: dashboard).
app.routing.ts
`import { Routes } from '@angular/router';

import { AdminLayoutComponent } from './layouts/admin/admin-layout.component';
import { AuthLayoutComponent } from './layouts/auth/auth-layout.component';
import { authenticatedGuard } from './guards/authenticatedGuard';

export const AppRoutes: Routes = [
    {
      path: '',
      redirectTo: 'pages/login',
      pathMatch: 'full',
    },{
      path: '',
      component: AdminLayoutComponent,
      canActivate:[authenticatedGuard],
      children: [
          {
              path: 'dashboard',
              loadChildren: './dashboard/dashboard.module#DashboardModule',
          }
      ]
    },{
      path: '',
      component: AuthLayoutComponent,
      children: [{
        path: 'pages',
        loadChildren: './pages/pages.module#PagesModule'
      }]
    }
];
`

authenticatedGuard
`import { Injectable } from '@angular/core';
import { Router, CanActivate, CanActivateChild, CanLoad } from '@angular/router';

@Injectable()

export class authenticatedGuard implements CanActivate, CanActivateChild, CanLoad {
	
	constructor(private routes:Router) { }
	canActivate():boolean{
		return this.checkLoginAuthentication();
	}
	canActivateChild():boolean{
		return this.checkLoginAuthentication();
	}
	canLoad():boolean{
		return this.checkLoginAuthentication();
	}
	checkLoginAuthentication():boolean{
		if(typeof(Storage) !='undefined'){
			let loginInfo = sessionStorage.getItem('info'),
				info;
			if(loginInfo){
				info = JSON.parse(loginInfo);
			}

			if(!(info && info._id)){
				this.routes.navigate(['/pages/login']);
			}
			return info && info._id;
		}
	}

}`
can you please tell me, what I am doing wrong here?

**Comment by bikramchoudhury** [comment]
is anyone there?

**Comment by alexandru-paduraru** [comment]
@bikramchoudhury yes, we are here. We've been looking for a solution but couldn't find it. We never used `authenitcationGuard` so we are still digging into this and we will come back with more details tomorrow.

Keep in touch,
Alex

**Comment by bikramchoudhury** [comment]
@alexandru-paduraru sure. Please update me, if you get any solution.

Thanks

**Comment by gassuncaocosta** [comment]
Hi, @bikramchoudhury!

I have implemented an AuthGuard and it's working fine. You must make a few changes to get it done.

I keept the App pointing to dashboard by default, instead login page.
First, the structure of my _**app.routing.ts**_ is almost the same as the original, just including the canActivate property:

```typescript
export const AppRoutes: Routes = [
    {
        path: '',
        redirectTo: 'dashboard',
        pathMatch: 'full',
    },
    {
        path: '',
        component: AdminLayoutComponent,
        canActivate: [AuthGuard], //added canActivate and AuthGuard service
        children: [
            {
                path: '',
                loadChildren: './dashboard/dashboard.module#DashboardModule'
            }
        ]
    },
    {
        path: '',
        component: AuthLayoutComponent,
        children: [{
            path: 'pages', //login page still under "pages" folder as the original code
            loadChildren: './pages/pages.module#PagesModule'
        }]
    }
];
```

Then, my _**auth-guard.service.ts**_ will redirect to original login page (under pages folder) if user is not logged in:

```typescript
export class AuthGuard implements CanActivate {
    constructor(private router: Router, private authService: AuthService) {}
    canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot):
        Observable<boolean> | Promise<boolean> | boolean {
        if (!this.authService.isLoggedIn()) {
            this.router.navigate(['pages/login'], { queryParams: { returnUrl: state.url } });
        }
        return true;
    }
}

/*
RouterStateSnapshot is optional. I'm using it for some cases wich user type a permanent link stored at browser history, but user is not logged in. For example:

User open browser an type: "http://localhost:4200/settings/system/preferences" but is not logged in.
Then the AuthGuard will redirects to login and, after login, the user will be redirected to the requested url, instead root page (dashboard).
*/
```

And finally, just for whim, if user enter login page after logged in, I send back to root (dashboard) in my _**pages/login/login.component.ts**_:

```typescript
constructor(private element: ElementRef, public authService: AuthService, 
    private router: Router, private route: ActivatedRoute) {
        if (this.authService.isLoggedIn()) {
            this.router.navigate(['dashboard']);
        } else {
            this.nativeElement = element.nativeElement;
            this.sidebarVisible = false;
        }
    }
```

Try this changes and let us know if helps you.

Regards.

**Comment by bikramchoudhury** [comment]
Thanks,
It works for me.

**Comment by IomarSantos** [comment]
I'm trying to implement with canLoad, but to no avail

**Comment by jperezrt** [comment]
Hi @gassuncaocosta , I have the same solution that you mentioned:

`this._loginService.loginWithEmail(email, pass)`
`   .then(() => this.router.navigate(['']))`
`            .catch(_error => {`
`                this.error = _error`
`                console.log( 'error:', this.error );`
`            });`

But I'm getting a loop error:

error: RangeError: Maximum call stack size exceeded
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:56)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)
    at LoginService.get [as isUserEmailLoggedIn] (login.service.ts:57)

can you help me with this please!!!

Thanks!
